src/java.base/share/classes/java/nio/file/Files.java

```diff
@@ -3351,7 +3352,14 @@ public static String readString(Path path, Charset cs) throws IOException {
         byte[] ba = readAllBytes(path);
         if (path.getClass().getModule() != Object.class.getModule())
             ba = ba.clone();
-        return JLA.newStringNoRepl(ba, cs);
+        try {
+            return JLA.newStringNoRepl(ba, cs);
+        } catch (IllegalArgumentException e) {
+            if (e.getCause() instanceof CharacterCodingException cause) {
+                throw cause;
+            }
+            throw e;
```

Comment:
This is wrong, if IAE is thrown for any other reason than CharacterCodingException then it means the JDK is broken.